gbn/gbn_conn.go

```diff
@@ -582,6 +788,13 @@ func (g *GoBackNConn) receivePacketsForever() error { // nolint:gocyclo
 			if gotValidACK {
 				g.resendTicker.Reset(g.resendTimeout)
 
+				g.awaitingCatchUpMu.RLock()
+				if m.Seq == g.awaitedACK && g.awaitingCatchUp {
+					log.Tracef("Got awaited ACK")
+					g.proceedAfterTime(g.catchUpId)
+				}
+				g.awaitingCatchUpMu.RUnlock()
```

Comment:
that should also help with the removal of the need for the catch up ID

Reply:
Commenting here regarding both this and (https://github.com/lightninglabs/lightning-node-connect/pull/87#discussion_r1389044721):

I've moved the logic into queue.ProcessACK instead!

1. Just to clarify, even though this does take an `RLock` before calling `proceedAfterTime`, it will not wait the entire timeout before releasing the `RLock`. This is since `time.AfterFunc` executes the callback in a separate `go-routine`, so the `RLock` will be released before the timeout has elapsed.

2. This doesn't remove the need for the `catchUpID`. Though I agree that maybe the `catchUpID` is overkill, and I can remove it if you think it just complicates more than it helps. I think I need to examplify with a scenario where the `catchUpID` is useful, to understand why I've added it. I should probably also add that I've experienced this while testing locally, so it does happen some times:
2a. Alice sends packets 1, 2 & 3 to Bob, but due latency she resend timeouts before receiving ACK 2 & 3 from Bob. Alice therefore resends packets 2 & 3.
2b. Alice now received the delayed ACKS 2 & 3 from Bob, and therefore executes `proceedAfterTime` for ACK 3 which starts the timeout before the it executes the callback. The callback will send the `awaitedACKSignal` once executed, if `awaitingCatchUp` is `true`.
2c. Alice now receives the NACK 4 that the resent packets in step (2a) led to. She now marks the resend catch up as done, and proceeds to send packets 4, 5 & 6.
2d. Due to packet loss, packet 5 never reaches Bob, so when Bob receives packet 6, he will respond with NACK 5.
2e. Alice now resends her queue again as the response to NACK 5, and then starts awaiting the catch up, and marks `awaitingCatchUp` as `true`.
2f. The callback from step `2b` now gets executed as the timeout has passed. As the the `awaitingCatchUp` is currently set to `true`, this callback will now mark the catch up incorrectly as done, and send the `awaitedACKSignal`. However, this callback that was executed was actually intended for the previous catch up cycle, and not the current cycle, so this callback shouldn't have marked the catch up as done.

This is what the `catchUpID` is used for. The current code instead also looks that the `catchUpID` is the still the same before marking the catch up as done, when a callback is eventually executed. Had that check existed for the above example, (2f) would not have marked the catch up as done.

Though I agree that it's maybe too complicated to understand why the `catchUpID` is needed. Its worth adding that the worst thing that likely can happen if we remove it, is that cases like the above would result in one extra resend cycle, so that's maybe not too bad compared to the complexity it adds.

Do you think I should still remove it or not, given the above explanation?

Reply:
> 2e. Alice now resends her queue again as the response to NACK 5, and then starts awaiting the catch up, and marks awaitingCatchUp as true.

Since Alice is currently busy with a resend/catch up, she should not re-trigger one here. 
I think the main thing leading to the added complexity is that the GBN is the one the the 'resendTimeout' and thus it looks like we are trying to let the queue work around this. How about instead letting the queue handle the when `resend` is allows to even be called/ or when the resent timeout ticks? 

I think a state machine approach would be best here. If we are in a resend/catch up state - then all parts should know this and act accordingly. 

Rather than adding complexity to fix this, I think we should first focus on reducing complexity 

Reply:
> Rather than adding complexity to fix this, I think we should first focus on reducing complexity 

I definitely agree with that we should attempt to reduce  complexity, and not add more complexity. And the `CatchUpID` was clearly a concept that was really complicated. I've now removed the `CatchUpID` and replaced this with an `caughtUpSignal` channel instead.
This new channel marks that we've reached the caught up state, so that any past delayed go routine knows that the resend cycle it was initiated in has already been completed, and that the goroutine therefore should be a no-op. I hope this reduces the complexity and makes the code more understandable. I hope this matches what you'd want, but let me know if you have an even more clean solution :).

> Since Alice is currently busy with a resend/catch up, she should not re-trigger one here.

The problem here is that Alice actually isn't busy with the resend/catch up before she re-triggers the resend in (2e). She correctly marked the resend catch up as completed in (2c) when she got the expected NACK 4. 

The issue is that even though she did receive the NACK 4 and marked the resend catch up as completed, the delayed callback that was initiated with the `time.AfterFunc` execution in `proceedAfterTime` in (2b) has not been executed yet, and will still be executed after the resendTimeout has passed. This is why we need to ensure that this delayed callback won't mess anything up once it finally executes, if we've entered a new resend catch up cycle once the delayed callback is finally executed.

If you know a way that would make it possible to stop any delayed callbacks that were initiated during a resend catch up cycle from actually executing, if the resend cycle has been completed, that would be a much better solution :)! Else I think this new `caughtUpSignal` channel is relatively easy to understand.

Reply:
just make the delayed call back check that the "state" of the catch up is still the same as when it was executed?

Reply:
This was now changed with the latest push to use a goroutine instead, so that the function is cancelled synchronously if it's still being executed when the resend catch up is over. Thanks for the help in simplifying this :)!